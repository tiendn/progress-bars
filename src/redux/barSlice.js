import { createSlice } from '@reduxjs/toolkit';
import { notification } from 'antd';

import { getBars } from '../apis/bars';

const defaultState = {
  buttons: [],
  bars: [],
  limit: 9999,
  currentBarIndex: 0,
  loading: true,
};

const barSlice = createSlice({
  name: 'bar',
  initialState: defaultState,
  reducers: {
    showLoading(state) {
      state.loading = true;
    },
    hideLoading(state) {
      state.loading = false;
    },
    fetchBarDataDone(state, action) {
      const { buttons, bars, limit } = action.payload;

      state.buttons = buttons || defaultState.buttons;
      state.bars = bars || defaultState.bars;
      state.limit = limit || defaultState.limit;
    },
    changeActiveBar(state, action) {
      state.currentBarIndex = action.payload.barIndex;
    },
    changeBarValue(state, action) {
      try {
        if (state.bars.length === 0) return;

        const { point } = action.payload;
        const { currentBarIndex, bars } = state;
        let newBarValue = bars[currentBarIndex] + point;

        if (newBarValue < 0) newBarValue = 0;
        state.bars[currentBarIndex] = newBarValue;
      } catch (err) {
        // console.log(err);
        // Report bug
      }
    },
  },
});

export const {
  changeActiveBar,
  changeBarValue,
  fetchBarDataDone,
  showLoading,
  hideLoading,
} = barSlice.actions;

export const getBarData = () => async dispatch => {
  try {
    dispatch(showLoading());
    const data = await getBars();
    dispatch(hideLoading());
    dispatch(fetchBarDataDone({ ...data }));
  } catch (err) {
    dispatch(hideLoading());
    notification.error({ message: err.message });
    console.log(err);
  }
};

export { defaultState };
export default barSlice.reducer;
