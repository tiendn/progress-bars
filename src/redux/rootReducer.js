import { combineReducers } from '@reduxjs/toolkit';
import barReducer from './barSlice';

const rootReducer = combineReducers({
  bar: barReducer,
});

export default rootReducer;
