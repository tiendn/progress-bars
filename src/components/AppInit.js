import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { Spin, Button, Result } from 'antd';

import { getBarData } from '../redux/barSlice';

const AppInit = ({ dispatch, loading }) => {
  useEffect(() => {
    dispatch(getBarData());
  }, []);

  const onRetry = () => {
    dispatch(getBarData());
  };

  if (loading)
    return (
      <div className="appContainer">
        <Spin size="large" data-testid="appLoading" />
      </div>
    );

  return (
    <div className="appDefaultContainer">
      <div data-testid="appDefault">
        <Result
          status="warning"
          title="There are some problems with your connection."
          extra={
            <Button
              type="primary"
              key="retry"
              style={{ width: 100 }}
              onClick={onRetry}
            >
              Retry
            </Button>
          }
        />
      </div>
    </div>
  );
};

const mapStateToProps = ({ bar }) => ({
  loading: bar.loading,
});

export default connect(mapStateToProps)(AppInit);
