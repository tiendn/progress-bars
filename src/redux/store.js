import { configureStore } from '@reduxjs/toolkit';
import { ThunkAction } from 'redux-thunk';
import rootReducer from './rootReducer';

const store = configureStore({
  reducer: rootReducer,
});

if (process.env.NODE_ENV === 'development' && module.hot) {
  module.hot.accept('./giphySlice', () => {
    const newRootReducer = require('./giphySlice').default;
    store.replaceReducer(newRootReducer);
  });
}

export type AppDispatch = typeof store.dispatch;
export type AppThunk = ThunkAction;
export default store;
