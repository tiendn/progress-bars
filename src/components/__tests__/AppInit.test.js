import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import AppInit from '../AppInit';
import configureStore from '../../redux/store';
import { hideLoading } from '../../redux/barSlice';

describe('AppInit testing', () => {
  let store;

  beforeAll(() => {
    store = configureStore;
  });

  it('renders with content', () => {
    const { baseElement } = render(
      <Provider store={store}>
        <AppInit />
      </Provider>,
    );

    expect(baseElement).toMatchSnapshot();
    expect(baseElement).not.toBeNull();
  });

  it('renders loading in default', async () => {
    const { getByTestId, container } = render(
      <Provider store={store}>
        <AppInit />
      </Provider>,
    );

    const Spin = await getByTestId('appLoading');
    expect(Spin).toMatchSnapshot();
    expect(container.contains(Spin)).toBe(true);
  });

  it('renders Result error when no data available', async () => {
    const { getByTestId, container } = render(
      <Provider store={store}>
        <AppInit />
      </Provider>,
    );

    store.dispatch(hideLoading());
    const ResultDefault = await getByTestId('appDefault');
    expect(ResultDefault).toMatchSnapshot();
    expect(container.contains(ResultDefault)).toBe(true);
  });
});
