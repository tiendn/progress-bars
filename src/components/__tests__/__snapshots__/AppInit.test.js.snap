// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppInit testing renders Result error when no data available 1`] = `
<div
  data-testid="appDefault"
>
  <div
    class="ant-result ant-result-warning"
  >
    <div
      class="ant-result-icon"
    >
      <i
        aria-label="icon: warning"
        class="anticon anticon-warning"
      >
        <svg
          aria-hidden="true"
          class=""
          data-icon="warning"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M955.7 856l-416-720c-6.2-10.7-16.9-16-27.7-16s-21.6 5.3-27.7 16l-416 720C56 877.4 71.4 904 96 904h832c24.6 0 40-26.6 27.7-48zM480 416c0-4.4 3.6-8 8-8h48c4.4 0 8 3.6 8 8v184c0 4.4-3.6 8-8 8h-48c-4.4 0-8-3.6-8-8V416zm32 352a48.01 48.01 0 0 1 0-96 48.01 48.01 0 0 1 0 96z"
          />
        </svg>
      </i>
    </div>
    <div
      class="ant-result-title"
    >
      There are some problems with your connection.
    </div>
    <div
      class="ant-result-extra"
    >
      <button
        class="ant-btn ant-btn-primary"
        style="width: 100px;"
        type="button"
      >
        <span>
          Retry
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`AppInit testing renders loading in default 1`] = `
<div
  class="ant-spin ant-spin-lg ant-spin-spinning"
  data-testid="appLoading"
>
  <span
    class="ant-spin-dot ant-spin-dot-spin"
  >
    <i
      class="ant-spin-dot-item"
    />
    <i
      class="ant-spin-dot-item"
    />
    <i
      class="ant-spin-dot-item"
    />
    <i
      class="ant-spin-dot-item"
    />
  </span>
</div>
`;

exports[`AppInit testing renders with content 1`] = `
<body>
  <div>
    <div
      class="appContainer"
    >
      <div
        class="ant-spin ant-spin-lg ant-spin-spinning"
        data-testid="appLoading"
      >
        <span
          class="ant-spin-dot ant-spin-dot-spin"
        >
          <i
            class="ant-spin-dot-item"
          />
          <i
            class="ant-spin-dot-item"
          />
          <i
            class="ant-spin-dot-item"
          />
          <i
            class="ant-spin-dot-item"
          />
        </span>
      </div>
    </div>
  </div>
</body>
`;
